sparkmagic/sparkmagic/controllerwidget/addendpointwidget.py

```diff
@@ -1,74 +1,84 @@
-                     constants.NO_AUTH: constants.NO_AUTH},
+        #map auth class path string to the instance of the class. 
+        self.auth_instances = {}
+        for auth in conf.authenticators().values(): 
+            module, class_name = (auth).rsplit('.', 1)
+            events_handler_module = importlib.import_module(module)
+            auth_class = getattr(events_handler_module, class_name)
+            self.auth_instances[auth] = auth_class()
+
+        self.auth_type = self.ipywidget_factory.get_dropdown(
+            options = conf.authenticators(),
             description=u"Auth type:"
         )
+        
+        #combine all authentication instance's widgets into one set to pass to self.children. 
+        self.all_widgets = list()
+        for _class, instance in self.auth_instances.items():
+            for widget in instance.widgets: 
+                if  (_class == self.auth_type.value): 
+                    widget.layout.display = 'flex'
+                    self.auth = instance
+                else: 
+                    widget.layout.display = 'none'
+                self.all_widgets.append(widget)
 
         # Submit widget
         self.submit_widget = self.ipywidget_factory.get_submit_button(
             description='Add endpoint'
         )
+ 
+        self.auth_type.on_trait_change(self._update_auth)
 
-        self.auth.on_trait_change(self._show_correct_endpoint_fields)
-
-        self.children = [self.ipywidget_factory.get_html(value="<br/>", width=widget_width),
-                         self.address_widget, self.auth, self.user_widget, self.password_widget,
-                         self.ipywidget_factory.get_html(value="<br/>", width=widget_width), self.submit_widget]
+        self.children = [self.ipywidget_factory.get_html(value="<br/>", width=WIDGET_WIDTH), self.auth_type] + self.all_widgets \
+        + [self.ipywidget_factory.get_html(value="<br/>", width=WIDGET_WIDTH),self.submit_widget ]
 
         for child in self.children:
             child.parent_widget = self
-
-        self._show_correct_endpoint_fields()
-
+        self._update_auth()
+        
     def run(self):
-        endpoint = Endpoint(self.address_widget.value, self.auth.value, self.user_widget.value, self.password_widget.value)
-        self.endpoints[self.address_widget.value] = endpoint
-        self.ipython_display.writeln("Added endpoint {}".format(self.address_widget.value))
-
-        # We need to call the refresh method because drop down in Tab 2 for endpoints wouldn't refresh with the new
-        # value otherwise.
-        self.refresh_method()
-
-    def _show_correct_endpoint_fields(self):
-        if self.auth.value == constants.NO_AUTH or self.auth.value == constants.AUTH_KERBEROS:
-            self.user_widget.layout.display = 'none'
-            self.password_widget.layout.display = 'none'
+        self.auth.update_with_widget_values()
+        if self.auth_type.label == "None":
```

Comment:
Is this special case still necessary? What if we just set self.auth to None if "None" is selected in the auth dropdown?

Reply:
I assume you mean when we are initializing the auth_instances map that maps the auth class path string to the instance of the class so "sparkmagic.auth.customauth.Authenticator" --> Authenticator() instead map  "sparkmagic.auth.customauth.Authenticator"  --> None. The issue with this is that later I iterate through all values in the auth_instances to build the all_widgets list (calling the get_widget() function for all the values  instance's widgets) and putting them all in one big list to add to self.children. So it would try to call None.get_widgets(), but the base authenticator class (what "None" maps to) actually has a widget (it is the address widget). The address widget is in the Authenticator and not the addwidgetendpoint.py because although most authenticators will use this address widget, putting it in authenticator class allows its subclasses the flexibility to override it (thats what google does). Also, None auth has a url widget because the behavior of sparkmagic shows the address widget when 'None' is selected.

Reply:
That sounds reasonable, thanks for clarifying.